library/serialize.c

```diff
@@ -0,0 +1,325 @@
+        var = ( expr );                         \
+        if( var < 0 )                           \
+        {                                       \
+            perror( #expr );                    \
+            goto cleanup;                       \
+        }                                       \
+    } while( 0 )
+static int mbedtls_serialize_prepare( void )
+{
+    int ret;
+    int host_to_target[2] = {-1, -1}, target_to_host[2] = {-1, -1};
+    const char *const frontend_path[] = {
+        "programs/host/frontend",
+        "../programs/host/frontend",
+        "../../programs/host/frontend",
+    };
+    const char *frontend_exe = NULL;
+    size_t i;
+
+    /* Look for frontend program */
+    for( i = 0; i < sizeof( frontend_path ) / sizeof( *frontend_path ); i++ )
+    {
+        if( access( frontend_path[i], F_OK ) >= 0 )
+        {
+            frontend_exe = frontend_path[i];
+            break;
+        }
+    }
+    if( frontend_exe == NULL )
+    {
+        fprintf( stderr, "Host frontend executable for offloading not found!\n" );
+        return( -1 );
+    }
+
+    /* Prepare the plumbing */
+    CHECK( pipe( host_to_target ), ret );
+    CHECK( pipe( target_to_host ), ret );
+    CHECK( signal( SIGTERM, relay_signal ) == SIG_ERR ? -1 : 0 , ret );
+    CHECK( fork( ), host_pid );
+
+    if( host_pid == 0 )
+    {
+        /* Child process (host */
+        char arg1[10], arg2[10];
+        sprintf( arg1, "%d", target_to_host[0] );
+        sprintf( arg2, "%d", host_to_target[1] );
+
+        /* Close peer ends of the pipe */
+        close( host_to_target[0] );
+        close( target_to_host[1] );
+
+        /* Pass pipe rd/wr descriptors to child */
+        execl( frontend_exe, frontend_exe, arg1, arg2, NULL );
+        perror( frontend_exe );
+        exit( 126 );
+    }
+
+    /* Parent process (target) */
+    serialize_read_fd = host_to_target[0];
+    close( host_to_target[1] );
+    close( target_to_host[0] );
+    serialize_write_fd = target_to_host[1];
+    atexit( kill_host_frontend );
+
+    return( 0 );
+cleanup:
+    if( host_to_target[0] >= 0 )
+        close( host_to_target[0] );
+    if( host_to_target[1] >= 0 )
+        close( host_to_target[1] );
+    if( target_to_host[0] >= 0 )
+        close( target_to_host[0] );
+    if( target_to_host[1] >= 0 )
+        close( target_to_host[1] );
+    return( -1 );
+}
+#undef CHECK
+
+static int mbedtls_serialize_write( const uint8_t *buffer, size_t length )
+{
+    ssize_t result;
+    if( serialize_write_fd == -1 )
+        if( mbedtls_serialize_prepare( ) != 0 )
+            return( MBEDTLS_ERR_SERIALIZE_SEND );
+    do {
+        result = write( serialize_write_fd, buffer, length );
+        /* This channel should never reach EOF under the current simplistic
+           design, so error out on EOF. */
+        if( result <= 0 )
+        {
+            return( MBEDTLS_ERR_SERIALIZE_SEND );
+        }
+        length -= result;
```

Comment:
Please double-check that the result of `write` is not larger than `length` to avoid underflow.

Reply:
This should not happen. `write()` writes up to length bytes from the buffer. It returns bytes written on success i.e. 1 - length. Or -1/0(EOF) on error. `result` can't be more than length.